llvm/lib/CAS/OnDiskHashMappedTrie.cpp

```diff
@@ -1236,6 +1383,21 @@ void HashMappedTrieHandle::print(
   return;
 }
 
+Error HashMappedTrieHandle::validate(
+    function_ref<Error(FileOffset, OnDiskHashMappedTrie::ConstValueProxy)>
+        RecordVerifier) const {
+  // Use the base TrieVisitor to identify the errors inside trie first.
+  TrieVisitor BasicVerifier(*this);
+  if (auto Err = BasicVerifier.visit())
+    return Err;
+
+  // If the trie data structure is sound, do a second pass to verify data and
+  // verifier function can assume the index is correct. However, there can be
```

Comment:
I don't understand why we do this in two passes instead of one.

Reply:
Rehashing requires reading the ref from the index so it would be good to validate all the index is correct first. Currently, the rehashing is just loading hash of the references using regular CAS API with no extra validation. The other motivation is that if it is a heavily connected CAS tree, it avoids validating the index multiple times, or requires to keep track of which trie node is validated (since the number can be very large).

Reply:
Of course, the current approach has the downside as noted in the comment.